logstash-core/src/main/java/org/logstash/util/CleanerThreadLocal.java

```diff
@@ -0,0 +1,115 @@
+package org.logstash.util;
+
+import java.lang.ref.Cleaner;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+/**
+ * A {@link CleanerThreadLocal} is semantically the same as a {@link ThreadLocal}, except that a clean action
+ * is called against each object when the {@link CleanerThreadLocal} no longer holds a reference to it.
+ * @param <T> the external value type
+ */
+public final class CleanerThreadLocal<T> {
```

Comment:
review note: the `CleanerThreadLocal` ended up not being necessary, and is removed in a later commit.